src/EditorFeatures/Core/Shared/Utilities/ThreadingContext.cs

```diff
@@ -49,5 +55,37 @@ public JoinableTaskFactory JoinableTaskFactory
         {
             get;
         }
+
+        public JoinableTaskCollection ShutdownBlockingTasks { get; }
+
+        private JoinableTaskFactory ShutdownBlockingTaskFactory { get; }
+
+        public CancellationToken DisposalToken => _disposalTokenSource.Token;
+
+        public JoinableTask RunWithShutdownBlockAsync(Func<CancellationToken, Task> func)
+        {
```

Comment:
Do you want to add `DisposalToken.ThrowIfCancellationRequested` before invoking their delegate? (assuming you keep this method after my other comments).
You'd want to update the docs on the method of course so callers know that may happen.
If you did this, you could shore up the documented guarantee on this method to say that if their delegate is ever invoked, it's guarantee to block shutdown.

Reply:
Our current usage matched this pattern, and it simplified the contract, so I applied this change.